server/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -730,10 +741,12 @@ private static ActionFilters setupActionFilters(List<ActionPlugin> actionPlugins
     public void initRestHandlers(Supplier<DiscoveryNodes> nodesInCluster) {
         List<AbstractCatAction> catActions = new ArrayList<>();
         Consumer<RestHandler> registerHandler = handler -> {
-            if (handler instanceof AbstractCatAction) {
-                catActions.add((AbstractCatAction) handler);
-            }
-            restController.registerHandler(handler);
+            if (shouldKeepRestHandler(handler)) {
+                if (handler instanceof AbstractCatAction) {
+                    catActions.add((AbstractCatAction) handler);
+                }
+                restController.registerHandler(handler);
+            } // else there's no way this handler can be reached, so no point in keeping it around
```

Comment:
What about having a dummy handler that returns a nicer error? That is, an error indicating this api is not available in serverless, instead of just "not found"?

Reply:
I wouldn't phrase it as "can't access", but rather it is unavailable in fully managed Elasticsearch. We want it to not exist, but having a more clear error message may help guide users. For example, if a user is looking at documentation for self managed Elasticsearch, and tries to follow an a example, getting a 404 (that they might have typo'd) could be frustrating. If instead they get a message like "this API is not available in fully managed Elasticsearch" it hints to them they need to check what they are doing (looking at self managed docs instead of fully managed docs).

Reply:
Alright the response now says something like `uri [/_cat/health] with method [GET] exists but is not available when running in serverless mode`.